### PR TITLE
compute_uv for SVD

### LINKED ISSUE #12420: [feature request] Add option for only computing the singular values of a matrix
## 🚀 Feature
I would like to propose the addition of an option to the current SVD implementation analogous to `compute_uv` in NumPy's SVD which will compute the singular values of a the matrix only.

## Pitch

```python
>>> torch.svd(a, compute_uv=False)
<tensor containing singular values>
>>> torch.svd(a, compute_uv=True)
<3-tuple standard output>
```

Methods that depend only on singular values such as `matrix_rank` and `nuclear_norm` will benefit from this addition.

### PR DESCRIPTION
**Pull request by vishwakftw**
Adds a `compute_uv` argument that defaults to `True` for optionally computing the singular vectors during SVD.

Closes https://github.com/pytorch/pytorch/issues/12420 .

### CODE REVIEW

File: tools/autograd/templates/Functions.cpp [hunk 1]
```diff
@@ -1489,7 +1489,10 @@ std::tuple<Tensor, Tensor, Tensor> prelu_double_backward(
 //
 // This makes no assumption on the signs of sigma.
 Tensor svd_backward(const std::vector<torch::autograd::Variable> &grads, const Tensor& self,
-          bool some, const Tensor& raw_u, const Tensor& sigma, const Tensor& raw_v) {
+          bool some, bool compute_uv, const Tensor& raw_u, const Tensor& sigma, const Tensor& raw_v) {
+  AT_CHECK(compute_uv,
+           "cannot compute backward for SVD without computing the singular matrices in the forward pass");
```
Comment (SsnL) [inline tools/autograd/templates/Functions.cpp]: Could you explicitly mention the `compute_uv` flag?

**Comment by SsnL** [comment]
I also wonder how much faster this is. Could you do some benchmarking on CPU (MKL) & GPU (MAGMA)? Thanks!

**Comment by vishwakftw** [comment]
CPU benchmarks:

size | compute_uv=True (master) | compute_uv=False (this PR)
----- | ----------------------- | ------------------------
100 x 200 | 1.88 ms ± 17.9 µs per loop | 1.11 ms ± 2.57 µs per loop
200 x 400 | 8.67 ms ± 499 µs per loop | 4.65 ms ± 591 µs per loop
500 x 250 | 13.6 ms ± 661 µs per loop | 7.1 ms ± 259 µs per loop
800 x 400 | 39.7 ms ± 2.72 ms per loop | 22.8 ms ± 1.21 ms per loop

GPU benchmarks can be expected to have the same speedup, since both use the divide and conquer algorithm (`gesdd`).

File: tools/autograd/templates/Functions.cpp [hunk 2]
```diff
@@ -1489,7 +1489,11 @@ std::tuple<Tensor, Tensor, Tensor> prelu_double_backward(
 //
 // This makes no assumption on the signs of sigma.
 Tensor svd_backward(const std::vector<torch::autograd::Variable> &grads, const Tensor& self,
-          bool some, const Tensor& raw_u, const Tensor& sigma, const Tensor& raw_v) {
+          bool some, bool compute_uv, const Tensor& raw_u, const Tensor& sigma, const Tensor& raw_v) {
+  AT_CHECK(compute_uv,
+           "Setting compute_uv to false doesn't compute singular matrices, and hence we cannot compute backward ",
```
Comment (zou3519) [inline tools/autograd/templates/Functions.cpp]: Could you compute the singular matrices here by calling svd?
Alternatively, could you make it more explicit that this is coming from svd_backward? something like "svd_backward: Setting compute_uv to false in torch.svd doesn't compute singular matrices, and hence we cannot compute backward. Please use torch.svd(compute_uv=True)"

Comment (SsnL) [reply tools/autograd/templates/Functions.cpp]: Yes I agree with Richard!

Comment (vishwakftw) [reply tools/autograd/templates/Functions.cpp]: Sure, I will modify the error message. I will change it for `symeig` as well.

**Comment by vishwakftw** [comment]
@SsnL Is this good to go now?

**Review by ssnl — APPROVED** [review]
LGTM, thanks!

**Pull request closed without merging**